tsl/src/nodes/gapfill/gapfill_plan.c

```diff
@@ -135,7 +135,17 @@ window_function_walker(Node *node, gapfill_walker_context *context)
 static bool
 gapfill_correct_order(PlannerInfo *root, Path *subpath, FuncExpr *func)
 {
-	if (list_length(subpath->pathkeys) != list_length(root->group_pathkeys))
+	List *groupby_pathkeys;
+#if PG16_LT
+	groupby_pathkeys = list_copy(root->group_pathkeys);
+#else
+	/* In PG16 group_pathkeys can contain additional pathkeys
+	 * used for optimization on ordered aggregates.
+	 * We only want to deal with group by elements only here.
+	 */
+	groupby_pathkeys = list_copy_head(root->group_pathkeys, root->num_groupby_pathkeys);
+#endif
+	if (list_length(subpath->pathkeys) != list_length(groupby_pathkeys))
 		return false;
```

Comment:
On PG16, the length condition is never false, do we need it?
Maybe we can remove it and use `list_copy_head` for all PG versions, for less code.
Or even better, rewrite the loop to use index, to avoid the copying and freeing altogether. In the current code, not all the return paths free the copy.

Reply:
Can't use the index everywhere because the `num_groupby_pathkeys` is present only in PG16.

I could rewrite it to use the index instead of copying the list, I would still need a check for versions so we can use `num_groupby_pathkeys` for PG16 and above, `list_length(root->group_pathkeys)` for all other.